### PR TITLE
Fix the bug that visible time is counted even after the page enters background

### LINKED ISSUE #7162: [amp-analytics] visible time is accumulated even when the page went to background
This is an issue in both visibility-v1 and visibility-v2.

@avimehta @dvoytenko

### PR DESCRIPTION
**Pull request by lannka**
Fixes #7162

### CODE REVIEW

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 1]
```diff
@@ -362,7 +365,12 @@ export class Visibility {
 
     if (!this.intersectionObserver_) {
       const onIntersectionChanges = entries => {
-        entries.forEach(this.onIntersectionChange_.bind(this));
+        entries.forEach(change => {
+          this.onIntersectionChange_(
+              change.target,
+              change.intersectionRatio * 100,
+              /* docVisible */true);
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: How do we know this is always `true`?

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 2]
```diff
@@ -393,47 +400,51 @@ export class Visibility {
...
     }
   }
 
-  /** @private */
-  onIntersectionChange_(change) {
-    const resource =
-        this.resourcesService_.getResourceForElement(change.target);
+  /**
+   * @param {!Element} target
+   * @param {number} visible
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Any better name for `visible` being a number?

File: extensions/amp-analytics/0.1/visibility-impl.js [hunk 3]
```diff
@@ -460,35 +467,24 @@ export class Visibility {
 
     // Remove target that have no listeners.
     if (listeners.length == 0) {
-      this.intersectionObserver_.unobserve(change.target);
+      this.intersectionObserver_.unobserve(target);
```
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: To double-check: `observe` is called for each trigger and `onubserve` is called only once after all triggers are done. Correct?

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 1 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/visibility-impl.js]: That's the nature of IntersectionObserver. It only trigger when doc visible.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 2 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/visibility-impl.js]: I was following the naming in V1. it makes sense to be consistent for now. we can rename after the removal of V1 code.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 3 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/visibility-impl.js]: yes. each trigger would yield a listener. and once we're done with all the listeners, we unobserve

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: What about hidden viewer case? I.e. `!viewer.isVisible()`?

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 3 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Listener per trigger is clear. But I'm wondering if `.observe` can be called many times like this.

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Ok

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 3 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/visibility-impl.js]: Yes, one observer can observe multiple targets

File: extensions/amp-analytics/0.1/visibility-impl.js (hunk 1 above)
Comment (lannka) [inline extensions/amp-analytics/0.1/visibility-impl.js]: that is handled by `onDocumentVisibilityChange_`

**Pull request merged**